src/libstd/io/mod.rs

```diff
@@ -0,0 +1,928 @@
+    ///
+    /// This function will only successfully return if an invocation of `Ok(0)`
+    /// succeeds at which point all bytes have been read into `buf`.
+    ///
+    /// # Errors
+    ///
+    /// If this function encounters an error of the kind
+    /// `ErrorKind::Interrupted` then the error is ignored and the operation
+    /// will continue.
+    ///
+    /// If any other read error is encountered then this function immediately
+    /// returns. Any bytes which have already been read will be present in
+    /// `buf` and the length will be adjusted appropriately.
+    fn read_to_end(&mut self, buf: &mut Vec<u8>) -> Result<()> {
+        read_to_end(self, buf)
+    }
+
+    /// Read all remaining bytes in this source, placing them into `buf`.
+    ///
+    /// # Errors
+    ///
+    /// If the data in this stream is *not* valid UTF-8 then an error is
+    /// returned and the length of `buf` is not adjusted.
+    ///
+    /// See `read_to_end` for other error semantics.
+    fn read_to_string(&mut self, buf: &mut String) -> Result<()> {
+        // Note that we do *not* call `.read_to_end()` here. We are passing
+        // `&mut Vec<u8>` (the raw contents of `buf`) into the `read_to_end`
+        // method to fill it up. An arbitrary implementation could overwrite the
+        // entire contents of the vector, not just append to it (which is what
+        // we are expecting).
+        //
+        // To prevent extraneously checking the UTF-8-ness of the entire buffer
+        // we pass it to our hardcoded `read_to_end` implementation which we
+        // know is guaranteed to only read data into the end of the buffer.
+        append_to_string(buf, |b| read_to_end(self, b))
+    }
+}
+
+/// Extension methods for all instances of `Read`, typically imported through
+/// `std::io::prelude::*`.
+pub trait ReadExt: Read + Sized {
+    /// Create a "by reference" adaptor for this instance of `Read`.
+    ///
+    /// The returned adaptor also implements `Read` and will simply borrow this
+    /// current reader.
+    fn by_ref(&mut self) -> &mut Self { self }
+
+    /// Transform this `Read` instance to an `Iterator` over its bytes.
+    ///
+    /// The returned type implements `Iterator` where the `Item` is `Result<u8,
+    /// R::Err>`.  The yielded item is `Ok` if a byte was successfully read and
+    /// `Err` otherwise for I/O errors. EOF is mapped to returning `None` for
+    /// this iterator.
+    fn bytes(self) -> Bytes<Self> {
+        Bytes { inner: self }
+    }
+
+    /// Transform this `Read` instance to an `Iterator` over `char`s.
+    ///
+    /// This adaptor will attempt to interpret this reader as an UTF-8 encoded
+    /// sequence of characters. The returned iterator will return `None` once
+    /// EOF is reached for this reader (and it's not in the middle of decoding a
+    /// character). Otherwise each element yielded will be a `Result<char, E>`
```

Comment:
So this will block indefinitely if we're reading a file that ends in an invalid utf8 sequence?